candidate-agreement/src/bft/accumulator.rs

```diff
@@ -0,0 +1,586 @@
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Message accumulator for each round of BFT consensus.
+
+use std::collections::{HashMap, HashSet};
+use std::collections::hash_map::Entry;
+use std::hash::Hash;
+
+use super::{Message, LocalizedMessage};
+
+/// Context necessary to execute a round of BFT.
+pub trait Context {
+	/// A full candidate.
+	type Candidate: Clone;
+	/// Unique digest of a proposed candidate (think hash).
+	type Digest: Hash + Eq + Clone;
+	/// Validator ID.
+	type ValidatorId: Hash + Eq;
+	/// A signature.
+	type Signature: Eq + Clone;
+}
+
+/// Justification at a given round.
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct Justification<D, S> {
+	/// The round.
+	pub round_number: usize,
+	/// The digest prepared for.
+	pub digest: D,
+	/// Signatures for the prepare messages.
+	pub signatures: Vec<S>,
+}
+
+impl<D, S> Justification<D, S> {
+	/// Fails if there are duplicate signatures or invalid.
+	///
+	/// Provide a closure for checking whether the signature is valid on a
+	/// digest.
+	///
+	/// The closure should return true iff the round number, digest, and signature
+	/// represent a valid prepare message and the signer was authorized to issue
+	/// it.
+	pub fn check<F, V>(&self, threshold: usize, check_message: F) -> bool
+		where
+			F: Fn(usize, &D, &S) -> Option<V>,
+			V: Hash + Eq,
+	{
+		let mut voted = HashSet::new();
+
+		let mut good = false;
+		for signature in &self.signatures {
+			match check_message(self.round_number, &self.digest, signature) {
+				None => return false,
+				Some(v) => {
+					if !voted.insert(v) {
+						return false;
+					} else if voted.len() >= threshold {
+						// don't return just yet since later signatures may be invalid.
+						good = true;
+					}
+				}
+			}
+		}
+
+		good
+	}
+}
+
+/// Type alias to represent a justification specifically for a prepare.
+pub type PrepareJustification<D, S> = Justification<D, S>;
+
+/// The round's state, based on imported messages.
+#[derive(PartialEq, Eq, Debug)]
+pub enum State<C, D, S> {
+	/// No proposal yet.
+	Begin,
+	/// Proposal received.
+	Proposed(C),
+	/// Seen 2f + 1 prepares for this digest.
+	Prepared(PrepareJustification<D, S>),
+	/// Seen 2f + 1 commits for a digest.
+	Committed(Justification<D, S>),
+	/// Seen 2f + 1 round-advancement messages.
+	Advanced(Option<PrepareJustification<D, S>>),
```

Comment:
what's the difference between `Advanced(None)` and `Begin`?

Reply:
`Begin` signifies that the proposer has not issued something yet. 

`Advanced(None)` might mean that due to network asynchrony, half of the validators have seen the proposal while half have not, and the result is that after a timeout all validators vote to advance.

It is possible that one validator has seen `Advanced(Some(x))` while others are at state `Advanced(None)`. Either way, we move onto the next proposer. Anyone honest who witnessed the threshold-prepare will propose that block again.